internal/pkg/types/v1alpha1/pack.go

```diff
@@ -0,0 +1,168 @@
+package v1alpha1
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/signal"
+	"strings"
+	"text/template"
+
+	"github.com/talos-systems/bldr/internal/pkg/config"
+	"github.com/talos-systems/bldr/internal/pkg/constants"
+	"github.com/talos-systems/gitmeta/pkg/git"
+	"github.com/talos-systems/gitmeta/pkg/metadata"
+	"golang.org/x/xerrors"
+)
+
+const tpl = `
+{{- $config := .Config -}}
+{{ $metadata := .Metadata -}}
+FROM {{ .Bldr }} AS build
+SHELL [ "{{ .Shell }}", "-c" ]
+{{ with .Install -}}
+{{ .AsDockerRun }}
+{{ end -}}{{ range $dep := .Dependencies -}}
+{{ $dep.AsDockerCopy $config $metadata }}
+{{ end -}}
+COPY . .
+RUN /bldr build
+
+{{ with .Finalize -}}
+FROM scratch
+{{ range $_, $f := . -}}
+COPY --from=build {{ $f.From }} {{ $f.To }}
+{{ end -}}
+{{ end -}}
+`
+
+func (p *Pkg) Pack() error {
+	c, err := config.Open()
+	if err != nil {
+		return err
+	}
+
+	if p.Variant == Scratch && p.Install != nil {
+		return xerrors.New("scratch variant does not support installs, you may want a dependency instead")
+	}
+
+	if p.Name == "" {
+		return xerrors.New("missing package name")
+	}
+
+	if p.Finalize == nil {
+		return xerrors.New("missing finalizer")
+	}
+
+	for _, f := range p.Finalize {
+		f.Normalize()
+	}
+
+	git, err := git.NewGit()
+	if err != nil {
+		return err
+	}
+	m, err := metadata.NewMetadata(git)
+	if err != nil {
+		return err
+	}
+
+	p.Config = c
+	p.Metadata = m
+
+	if p.Bldr == "" {
+		p.Bldr = fmt.Sprintf("%s/%s/%s:%s", p.Config.Repository, p.Config.Username, "bldr", constants.Version)
+	}
+
+	p.Bldr = p.Bldr + "-" + p.Variant.String()
+
+	dockerfile, err := render(p)
+	if err != nil {
+		return err
+	}
+
+	tmpfile, err := os.Create(".pkg.dockerfile")
+	if err != nil {
+		return err
+	}
+
+	defer os.Remove(tmpfile.Name())
+
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+	go func() {
+		<-interrupt
+		os.Remove(tmpfile.Name())
+	}()
+
+	if _, err := tmpfile.Write([]byte(dockerfile)); err != nil {
+		return err
+	}
+	if err := tmpfile.Close(); err != nil {
+		return err
+	}
+
+	args := []string{
+		"buildx",
+		"build",
+		"--platform=" + getenv("PLATFORM", "linux/amd64,linux/arm64,linux/arm/7"),
+		"--progress=" + getenv("PROGRESS", "auto"),
+		"--push=" + getenv("PUSH", push(p.Metadata)),
+		"--cache-from=" + getenv("CACHE_FROM", strings.Join([]string{p.Config.Repository, p.Config.Username, p.Name}, "/")+":cache"),
+		"--cache-to=" + getenv("CACHE_TO", strings.Join([]string{p.Config.Repository, p.Config.Username, p.Name}, "/")+":cache"),
+		"--tag=" + strings.Join([]string{p.Config.Repository, p.Config.Username, p.Name}, "/") + ":" + p.Metadata.Container.Image.Tag,
+		"--file=" + tmpfile.Name(),
+		".",
+	}
+
+	return cmd("docker", args...)
+}
+
+func (i Install) AsDockerRun() string {
+	return fmt.Sprintf("RUN apk add --no-cache %s", strings.Join(i, " "))
```

Comment:
fine for now but probably want to add a switch statement based on base image ( idk if regex or strings.contains )
```
switch 
case "alpine"
case "ubuntu"
case "centos"
...
```

Reply:
Absolutely. Cool if we add this on later? I agree 100%, just have my concerns with time.